##R script to run PCADAPT and make plots of PCA results and manhattan plots

#Libraries
library(pcadapt)
library(qvalue)
library(qqman)
library(ggplot2)

#################################################### Load files ####################################################

##### Load data files that will be used for making plots with results 
####  Require files for info (.fam, .map, and pop coding info files)

#Individual information file (.fam format from PLINK)
fam=read.table("~/Dropbox/Desktop/postdocwork/Salmon/GenomicVulnerability/PCAdapt/NorthAM_86pops_goodloc_maf005.fam", header=F)
#Population codes will be created from this with the "ID" name (removing everything after underscore _)

#SNP info file (.map format from PLINK)
map=read.table("~/Dropbox/Desktop/postdocwork/Salmon/GenomicVulnerability/PCAdapt/NorthAM_86pops_goodloc_maf005.bim", header=F)

#Region info for plotting:
regions=read.table("~/Dropbox/Desktop/postdocwork/Salmon/GenomicVulnerability/NorthAM_Locations_220K_with_Region.txt", header=T)
#In regions file - pop IDs should be in first column

#################################################### PCAdapt script ####################################################

#Run PCAdapt - use .bed format file from PLINK
filename <- read.pcadapt("~/Dropbox/Desktop/postdocwork/Salmon/GenomicVulnerability/PCAdapt/NorthAM_86pops_goodloc_maf005.bed",  type = "bed")
x <- pcadapt(filename, method = "mahalanobis", K=2)

#Combine results (x) from pcadapt with individual data (fam file)
combined=as.data.frame(cbind(x$scores, fam[,1:2]))
colnames(combined)=c("PC1", "PC2", "Pop", "ID")

#Remove extra text in ID names to create correct "Pop2" Column
combined$Pop2=gsub(combined$ID, pattern="\\_.*", replacement = "")

#Combine information about regions with pcadapt results
all_data_for_plotting=merge(x=combined, by.x="Pop2", y=regions, by.y=1)

#Get variance explained for axes
variance_explained=(x$singular.values)^2

#Colour for points (used in paper)
col_scheme_paper=c("#ff7f00", #LAB
                   "#1f78b4", #Maine
                   "#ffff33", #NB
                   "darkorchid1", #NL
                   "#33a02c", #NS
                   "#e31a1c", #PEI
                   "black") #QC

#GGplot of PC1 and PC2
ggplot(all_data_for_plotting, aes(x=PC1, y=PC2))+ #Use PC1 and PC2 data
  theme_bw()+ #limited theme
  geom_point(col="gray30",aes(fill=Region), size=3,pch=21)+ #Colour points by region
  scale_fill_manual(values=col_scheme_paper)+ #Colours used for paper
  theme(legend.position = "top", 
        axis.title =  element_text(size=20),
        axis.text = element_text(size=13))+
  xlab(paste0("PC1 (", variance_explained[1], "%)")) +  #Add axis titles with percent variance explained based on PCAdapt
  ylab(paste0("PC2 (", variance_explained[2], "%)")) 
#Save as 10 x 10 


#Save data 
#write.table(all_data_for_plotting, "~/Desktop/Salmon_GenomicVulnerability/PCAdapt/PCAResults_PC1_PC2_inds.txt", quote = F, row.names = F, col.names = T, sep="\t")


#################################################### Manhattan plots ####################################################

#Combine SNP info (map file with position information) with results (p-values from pcadapt)
#Both datasets are in the same order (output from PLINK) - so can use cbind() 
snp_info=as.data.frame(cbind(map, x$pvalues))

#Get q-value transformation
qval_transform=qvalue(snp_info$`x$pvalues`)
snp_info$qval=qval_transform$qvalues #Add q-val column

#Add column names
colnames(snp_info)=c("Chr", "SNP", "x", "Pos", "a", "b", "pval", "qval") 

#Save all pcadapt results - p/q values for each SNP with position info 
write.table(snp_info, "Save_File.txt", quote = F, row.names = F, col.names = T, sep="\t")


#Plot all chromosomes together
#Check ylim if need to increase value of maximum -log10 p value
manhattan(x=snp_info[!is.na(snp_info$Chr),], chr = "Chr",
          bp = "Pos", snp = "SNP",p = "qval", suggestiveline = FALSE, ylim=c(0,100),
          genomewideline = -log10(0.05), ylab="-log10(q)", cex=0.9,
          col = c("gray30", "dodgerblue2")) 



####Used this layout to plot multiple panels together in qqman (uses base R)
layout(matrix(c(1,1,2,1,1,3, 1,1,4), 3, 3, byrow = TRUE))

manhattan(x=snp_info[!is.na(snp_info$Chr),], chr = "Chr",
          bp = "Pos", snp = "SNP",p = "qval", suggestiveline = FALSE, ylim=c(0,100),
          genomewideline = -log10(0.05), ylab="-log10(q)", cex=0.9,
          col = c("gray30", "dodgerblue2")) 

#These have been editted to show areas of interest (added text to point out genes)
manhattan(x=snp_info[which(snp_info$Chr==9),], chr = "Chr",
          bp = "Pos", snp = "SNP",p = "qval", suggestiveline = FALSE, 
          genomewideline = -log10(0.05), ylab="-log10(q)",
          col = c("gray30", "dodgerblue2"), ylim=c(0, 11))
text(x=24905552+8e6, y=10, label="six-6", col="black" ) #six6 end
snp_info[which(snp_info$Chr==9 & snp_info$qval < 1e-8),]

manhattan(x=snp_info[which(snp_info$Chr==14),], chr = "Chr",
          bp = "Pos", snp = "SNP",p = "qval", suggestiveline = FALSE, 
          genomewideline = -log10(0.05), ylab="-log10(q)",
          col = c( "dodgerblue2"), ylim=c(0, 12))
snp_info[which(snp_info$Chr==14 & snp_info$qval < 1e-10),]
text(x=71688280+1e7, y=11, label="MHC1-l \n (pseudo)", col="black") #MHC end


manhattan(x=snp_info[which(snp_info$Chr==19),], chr = "Chr",
          bp = "Pos", snp = "SNP",p = "qval", suggestiveline = FALSE, 
          genomewideline = -log10(0.05), ylab="-log10(q)",
          col = c("gray30", "dodgerblue2") )
text(x=77173861 -1e7, y=8.5,label="Leukocyte \n elastase \n inhibitor")

#Saved 4 panel as 16x9

